web/recv.js

```diff
@@ -0,0 +1,401 @@
+
+  var _video = 0;
+  var _workers = [];
+  var _nextWorker = 0;
+  var _workerReady;
+  var _supportedFormats = ["NV12", "I420"]; // have cimbard_* return this somehow?
+
+  function _toggleFullscreen() {
+    if (document.fullscreenElement) {
+      return document.exitFullscreen();
+    }
+    else {
+      return document.documentElement.requestFullscreen();
+    }
+  }
+
+  function _downloadHelper(name, bloburl) {
+    var aaa = document.createElement('a');
+    aaa.href = bloburl;
+    aaa.download = name;
+    document.body.appendChild(aaa);
+    aaa.style = 'display: none';
+    aaa.click();
+    aaa.remove();
+  }
+
+  // public interface
+  return {
+    init: function (video, num_workers) {
+      Recv.init_ww(num_workers);
+      Recv.init_video(video);
+    },
+
+    set_error: function (msg) {
+      Recv.set_HTML('errorbox', msg);
+      return false;
+    },
+
+    ww_ready: new Promise(resolve => {
+      _workerReady = resolve;
+    }),
+
+    init_ww: function (num_workers) {
+      // clean up _workers if exists?
+      _workers = [];
+      for (let i = 0; i < num_workers; i++) {
+        _workers.push(new Worker('recv-worker.js'));
+
+        _workers[i].onmessage = (event) => {
+          Recv.on_decode(i, event.data);
+        };
+
+        _workers[i].onerror = (error) => {
+          console.error('Worker' + i + ' error:', error);
+        };
+      }
+    },
+
+    init_video: function (video) {
+      _video = video;
+
+      var constraints = {
+        audio: false,
+        video: {
+          width: { min: 720, ideal: 1920 }, // Request HD but allow flexibility
+          height: { min: 720, ideal: 1080 },
+          aspectRatio: matchMedia('all and (orientation:landscape)').matches ? 16 / 9 : 9 / 16,
+          facingMode: 'environment',
+          exposureMode: 'continuous',
+          focusMode: 'continuous',
+          frameRate: { ideal: 15, max: 30 }, // we're not trying to set the user's phone on fire
+        }
+      };
+
+      if (!navigator.mediaDevices || !navigator.mediaDevices.getUserMedia) {
+        return Recv.set_error('mediaDevices not supported? :(');
+      }
+
+      navigator.mediaDevices.getUserMedia(constraints)
+        .then(localMediaStream => {
+          //console.log(localMediaStream);
+          //console.dir(video);
+          if ('srcObject' in video) {
+            video.srcObject = localMediaStream;
+          } else {
+            video.src = URL.createObjectURL(localMediaStream); //deprecated
+          }
+          video.play();
+          video.requestVideoFrameCallback(Recv.on_frame);
+        })
+        .catch(err => {
+          console.error(`OH NO!!!!`, err);
+        });
+    },
+
+    restart_video: function () {
+      console.log("restart video?");
+      if (_video) {
+        return;
+      }
+      if (!('srcObject' in _video)) {
+        console.log("can't restart video? :|")
+        return;
+      }
+      // ios cancels the feed sometimes. Restart it...?
```

Comment:
I mentioned this above, but I don't think we necessarily need this anymore. I think this case may have been due to an unnecessary `alert()`